pkgs/native_assets_builder/doc/protocol/v1.md

```diff
@@ -0,0 +1,227 @@
+Build Hook Specification
+========================
+
+Version 1: As of 2024-01-24 implemented behind `--enable-experiment=native-assets`
+
+### Concepts
+
+#### Asset
+
+An asset is data which is accessible from a Dart or Flutter application. To retrieve an asset at runtime, we associate with it a string as a unique identifier, the `assetId`. There are several example types of assets:
+* Assets which designate symbols present in the target system, process, or executable. They are identified by their name.
+* Dynamic libraries bundled into the application.
+
+An application is compiled to run on a certain target OS and architecture. If different targets require different assets, the package developer must specify which asset to bundle for which target.
+
+An asset has different ways of being accessible in the final application. It is either brought in "manually" by having the package developer specify a path of the asset on the current system, it can be part of the Dart or Flutter SDK, or it can be already present in the target system. If the asset is bundled "manually", the Dart or Flutter SDK will take care of copying the asset from its specified location on the current system into the application bundle.
+
+Assets are also called "native assets" to differentiate them from the Dart code also bundled with an application.
+
+#### AssetId
+An asset must have a string identifier called `assetId`. Dart code that uses an asset, references the asset using this `assetId`.
+
+A package must prefix all `assetId`s it defines with: `package:<package>/`, `<package>` being the current package's name. This ensures assets don't conflict between packages.
+
+Additionally, the convention is that an asset referenced from `lib/src/foo.dart` in `package:foo` has the `assetId` `'package:foo/src/foo.dart'`.
+
+#### Target
+
+A target specifies the operating system and architecture of the targeted platform of the application.
+
+#### `build.dart`
+If there are assets in a package which have to be "manually" bundled, i.e. they are not shipped as part of Dart or Flutter and are also not present on the target system by default, that package must define a `build.dart` script specifying these assets. This script takes as input information such as the target OS and architecture in a `build_config.yaml` file, and returns as output a list of assets to be bundled in a `build_output.yaml` file. 
```

Comment:
Also the assets which are on the system or in Dart/Flutter must be declared in `build.dart` in the `build_output.yaml`. However, `build.dart` will not write a file for them to the output dir.